### ISSUE TITLE
How to update the params in batchnorm layers by passing the inputs

### ISSUE THREAD

**Issue by fryng** [keras-team/keras#10533]
I'm currently implementing Averaging Weights Leads to Wider Optima and Better Generalization in keras ,the source code(Pytorch) is  [here](https://github.com/fastai/fastai/blob/master/fastai/swa.py) , but i'm stucking in the **fix_batchnorm()** function.  **how could i update the params(moving_mean,moving_variance) in batchnorm layers by passing the inputs?**  In Pytorch, u  just do like this `model(inputs)`, In keras , `model.predict(inputs)` doesn't work
the pytorch code is here
```
def fix_batchnorm(swa_model, train_dl):
    """
    During training, batch norm layers keep track of a running mean and
    variance of the previous layer's activations. Because the parameters
    of the SWA model are computed as the average of other models' parameters,
    the SWA model never sees the training data itself, and therefore has no
    opportunity to compute the correct batch norm statistics. Before performing 
    inference with the SWA model, we perform a single pass over the training data
    to calculate an accurate running mean and variance for each batch norm layer.
    """
    bn_modules = []
    swa_model.apply(lambda module: collect_bn_modules(module, bn_modules))
    
    if not bn_modules: return

    swa_model.train()

    for module in bn_modules:
        module.running_mean = torch.zeros_like(module.running_mean)
        module.running_var = torch.ones_like(module.running_var)
    
    momenta = [m.momentum for m in bn_modules]

    inputs_seen = 0

    for (*x,y) in iter(train_dl):        
        xs = V(x)
        batch_size = xs[0].size(0)

        momentum = batch_size / (inputs_seen + batch_size)
        for module in bn_modules:
            module.momentum = momentum
                            
        res = swa_model(*xs)        
        
        inputs_seen += batch_size
                
    for module, momentum in zip(bn_modules, momenta):
        module.momentum = momentum    
```
i do like this:
```
def fix_batchnorm(swa_model, generater):
    bn_layers = []

    for swa_layer in zip(swa_model.layers):
        if isinstance(swa_layer, keras.layers.BatchNormalization):
            bn_layers.append(swa_layer)

    if not bn_layers: return
    
    for layer in bn_layers:
        layer.moving_mean = K.zeros_like(layer.moving_mean)
        layer.moving_variance = K.zeros_like(layer.moving_variance)
    momenta = [l.momentum  for l in bn_layers]

    inputs_seen = 0

    for X_batch , Y_batch in generater:
        batch_size = X_batch.shape[0]

        momentum = batch_size / (inputs_seen + batch_size)
        for layer in bn_layers:
            layer.momentum = momentum
       
        """===========Im stucking here==============="""
        swa_model.predict(X_batch)
        """====================================="""

        inputs_seen += batch_size 

    for layer, momentum in zip(bn_layers, momenta):
        layer.momentum = momentum 
```

**Comment by spsancti** [comment]
Well, found it.
You have to save model and then do:
```
    K.clear_session() 
    K.set_learning_phase(1) 
    model = load_model("...")
```
And then your batchnorm will update with predict or evaluate.

**Comment by AITTSMD** [comment]
@spsancti Hi, I try your method, but it doesn't work. Running mean and var stay same.
